tiny_cnn/layers/dropout_layer.h

```diff
@@ -134,22 +142,26 @@ class dropout_layer : public layer {
 
     std::string layer_type() const override { return "dropout"; }
 
-    const std::vector<uint8_t>& get_mask(cnn_size_t worker_index) const {
-        return dropout_layer_worker_storage_[worker_index].mask_;
+    const std::vector<uint8_t>& get_mask(cnn_size_t worker_index, cnn_size_t sample_index) const {
+        return dropout_layer_worker_storage_[worker_index].mask_[sample_index];
     }
 
     virtual void set_worker_count(cnn_size_t worker_count) override {
         Base::set_worker_count(worker_count);
         dropout_layer_worker_storage_.resize(worker_count);
 
         for (dropout_layer_worker_specific_storage& dws : dropout_layer_worker_storage_) {
-            dws.mask_.resize(in_size_);
+            if (dws.mask_.empty()) {
+                dws.mask_.resize(1, std::vector<uint8_t>(in_size_));
```

Comment:
@nyanp: Correct. For other layer types I took the latter approach, though (see e.g. [here](https://github.com/reunanen/tiny-cnn/blob/80272e443b10c3f8ed53d6d28da054d38aee00b8/tiny_cnn/layers/convolutional_layer.h#L494-L495) – although I think this may be one of the places that are not correct yet, but let's not focus on that right now).

Not sure why it requires a lock, though, as each vector (to be resized) is still worker-specific.